### PR TITLE
feat: switch to pydantic

### PR DESCRIPTION
**Pull request by PurpleMyst**

### CODE REVIEW

**Comment by PurpleMyst** [comment]
@Akuli I managed to fix the FIXME type comment in this PR, but I'm not actually sure how to test this all using porcupine. Could you do so for me or delineate how to do so please?

**Comment by Akuli** [comment]
I tested this with porcupine like this, in porcupine with its venv activate:

```
$ pip uninstall sansio_lsp_client
$ pip install https://github.com/PurpleMyst/sansio-lsp-client/archive/pydantic-rewrite.zip
$ python3 -m porcupine release.py
```

By default, Porcupine is configured to work with `python-language-server`, and to test that, I opened `release.py` which is a Python script that Porcupine comes with. It doesn't work:

```
log file: /home/akuli/.cache/porcupine/logs/2021-01-13T10-27-18.txt
porcupine.plugins.langserver.4154 ERROR: error while receiving lsp events
Traceback (most recent call last):
  File "/home/akuli/porcu/porcupine/plugins/langserver.py", line 376, in _run_stuff_once
    lsp_events = self._lsp_client.recv(received_bytes)
  File "/home/akuli/porcu/env/lib/python3.7/site-packages/sansio_lsp_client/client.py", line 118, in recv
    messages = list(_parse_messages(self._recv_buf))
  File "/home/akuli/porcu/env/lib/python3.7/site-packages/sansio_lsp_client/io_handler.py", line 187, in _parse_messages
    parsed = _parse_one_message(response_buf)
  File "/home/akuli/porcu/env/lib/python3.7/site-packages/sansio_lsp_client/io_handler.py", line 180, in _parse_one_message
    return [do_it(content)]
  File "/home/akuli/porcu/env/lib/python3.7/site-packages/sansio_lsp_client/io_handler.py", line 160, in do_it
    request = Request.parse_obj(data)
  File "pydantic/main.py", line 520, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 362, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Request
method
  field required (type=value_error.missing)
```

File: sansio_lsp_client/events.py [hunk 1]
```diff
@@ -1,90 +1,80 @@
...
     Diagnostic,
-    Range,
     MessageType,
     MessageActionItem,
     CompletionItem,
     CompletionList,
     TextEdit,
 )
 
+Id = t.Union[int, str]
```
Comment (Akuli) [inline sansio_lsp_client/events.py:17]: When is this a string?

Comment (PurpleMyst) [reply sansio_lsp_client/events.py:17]: I'm not sure if it ever *actually* is, but the spec defines it as `id: integer | string;`

**Comment by PurpleMyst** [comment]
@Akuli I've fixed the "this does not start up at all" bug, the issue was depending on TypeError while ValidationError was actually a subclass of ValueError. Anyways, could you please test again? It seems to start up for me.

**Comment by Akuli** [comment]
Ideally, there would be a way to merge #16 and #17 so that you could use pytest for doing the same things that Porcupine does when I test with it. I tried that, and it resulted in a merge conflict in poetry.lock. I tried deleting poetry.lock to resolve it, which makes the merge work but then `poetry install` fails:

```
$ poetry install
Updating dependencies
Resolving dependencies... (10.6s)

Writing lock file

Package operations: 1 install, 4 updates, 0 removals

  • Updating typed-ast (1.4.1 -> 1.4.2)
  • Updating typing-extensions (3.7.4.2 -> 3.7.4.3)
  • Updating toml (0.10.1 -> 0.10.2)
  • Updating attrs (18.2.0 -> 20.3.0)
  • Installing pydantic (1.7.3): Failed

  EnvCommandError

  Command ['/home/akuli/sansio-lsp-client/env/bin/pip', 'install', '--no-deps', '/home/akuli/.cache/pypoetry/artifacts/a2/97/d0/ea9e192cb9618b7c2e414860077a8353f5153e377eab0aba2e3844f50e/pydantic-1.7.3-cp37-cp37m-manylinux2014_x86_64.whl'] errored with the following return code 1, and output: 
  pydantic-1.7.3-cp37-cp37m-manylinux2014_x86_64.whl is not a supported wheel on this platform.
  

  at env/lib/python3.7/site-packages/poetry/utils/env.py:1074 in _run
      1070│                 output = subprocess.check_output(
      1071│                     cmd, stderr=subprocess.STDOUT, **kwargs
      1072│                 )
      1073│         except CalledProcessError as e:
    → 1074│             raise EnvCommandError(e, input=input_)
      1075│ 
      1076│         return decode(output)
      1077│ 
      1078│     def execute(self, bin, *args, **kwargs):
```

I fucking hate poetry...

Anyway, I then installed the dependencies with `pip`, just looking them up from `pyproject.toml`:

```
$ pip install dataclasses pydantic pytest mypy black python-language-server
```

Then the tests passed. I'll try with Porcupine next, but I would expect it to work with Porcupine too.

**Comment by Akuli** [comment]
It works

**Review by Akuli — APPROVED** [review]

**Comment by Akuli** [comment]
Consider updating mypy to get rid of problems like 30e7c18585

**Comment by PurpleMyst** [comment]
I've merged #17 and I've rebased this branch on those changes. I think your poetry version's outdated like my mypy was :D

**Comment by PurpleMyst** [comment]
@Akuli the tests seem not to pass?

**Comment by PurpleMyst** [comment]
Woops, I'd forgotten to include the fix commit when rebasing and hard-resetting. Thankfully `git reflog` saved me. Does this all look good to you? Can I merge?

**Review by Akuli — APPROVED** [review]

**Comment by Akuli** [comment]
Can be merged when mypy and black pass

**Comment by Akuli** [comment]
mypy-update branch contains newer version of mypy added to poetry. It doesn't prevent the weird `object` error from happening though...

**Comment by PurpleMyst** [comment]
Yeah, I'm not sure why this error happens. For all intents and purposes, `t.Union[...]` should be a `t.Type`. I'm just going to commit the `# type: ignore[arg-type]` comment again

**Comment by PurpleMyst** [comment]
oh my god this interaction between black and mypy is killing me.

**Comment by Akuli** [comment]
yay :tada:

**Comment by PurpleMyst** [comment]
There you go. Can I squash&merge?

**Pull request merged**